UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapDxeSmm.c

```diff
@@ -0,0 +1,715 @@
+          DEBUG ((
+            DEBUG_INFO,
+            "UefiMemory protection: 0x%lx - 0x%lx %r\n",
+            MemoryMap->PhysicalStart,
+            MemoryMap->PhysicalStart + (UINT64)EFI_PAGES_TO_SIZE ((UINTN)MemoryMap->NumberOfPages),
+            Status
+            ));
+        }
+
+        MemoryMap = NEXT_MEMORY_DESCRIPTOR (MemoryMap, mUefiDescriptorSize);
+      }
+    }
+
+    //
+    // Do not free mUefiMemoryMap, it will be checked in IsSmmCommBufferForbiddenAddress().
+    //
+
+    //
+    // Set untested memory as not present.
+    //
+    if (mGcdMemSpace != NULL) {
+      for (Index = 0; Index < mGcdMemNumberOfDesc; Index++) {
+        Status = SmmSetMemoryAttributes (
+                   mGcdMemSpace[Index].BaseAddress,
+                   mGcdMemSpace[Index].Length,
+                   EFI_MEMORY_RP
+                   );
+        DEBUG ((
+          DEBUG_INFO,
+          "GcdMemory protection: 0x%lx - 0x%lx %r\n",
+          mGcdMemSpace[Index].BaseAddress,
+          mGcdMemSpace[Index].BaseAddress + mGcdMemSpace[Index].Length,
+          Status
+          ));
+      }
+    }
+
+    //
+    // Do not free mGcdMemSpace, it will be checked in IsSmmCommBufferForbiddenAddress().
+    //
+
+    //
+    // Set UEFI runtime memory with EFI_MEMORY_RO as not present.
+    //
+    if (mUefiMemoryAttributesTable != NULL) {
+      Entry = (EFI_MEMORY_DESCRIPTOR *)(mUefiMemoryAttributesTable + 1);
+      for (Index = 0; Index < mUefiMemoryAttributesTable->NumberOfEntries; Index++) {
+        if ((Entry->Type == EfiRuntimeServicesCode) || (Entry->Type == EfiRuntimeServicesData)) {
+          if ((Entry->Attribute & EFI_MEMORY_RO) != 0) {
+            Status = SmmSetMemoryAttributes (
+                       Entry->PhysicalStart,
+                       EFI_PAGES_TO_SIZE ((UINTN)Entry->NumberOfPages),
+                       EFI_MEMORY_RP
+                       );
+            DEBUG ((
+              DEBUG_INFO,
+              "UefiMemoryAttribute protection: 0x%lx - 0x%lx %r\n",
+              Entry->PhysicalStart,
+              Entry->PhysicalStart + (UINT64)EFI_PAGES_TO_SIZE ((UINTN)Entry->NumberOfPages),
+              Status
+              ));
+          }
+        }
+
+        Entry = NEXT_MEMORY_DESCRIPTOR (Entry, mUefiMemoryAttributesTable->DescriptorSize);
+      }
+    }
+
+    //
+    // Do not free mUefiMemoryAttributesTable, it will be checked in IsSmmCommBufferForbiddenAddress().
+    //
+  }
+
+  //
+  // Set execute-disable flag
+  //
+  mXdEnabled = TRUE;
+
+  WRITE_PROTECT_RO_PAGES (WriteProtect, CetEnabled);
+
+  DEBUG ((DEBUG_INFO, "UpdateUefiMemMapAttributes Done.\n"));
+}
+
+/**
+  Check SmmProfile is enabled or not.
+
+  @param[in, out]     BaseAddress     Return BaseAddress of SmmProfileData.
+  @param[in, out]     Size            Return Size of SmmProfileData.
+
+  @return TRUE     SmmProfile is enabled.
+          FALSE    SmmProfile is not enabled.
+
+**/
+BOOLEAN
+IsSmmProfileEnabled (

```

Comment:
How about seperate the IsSmmProfileEnabled() into two different function for the legacy smm？Acctually it does two different things:
1.Retrive if smm profile is enabled and the smm profile base and length
2.Allocate smm profile range if enabled and Input parameters are valid